### PR TITLE
Wc/snap to terrain - CsvSpawner: Make "snap to terrain" based on collision layer.

### LINKED ISSUE #70: CsvSpawner: Make "snap to terrain" based on collision layer.
Currently "Place on terrain" parameter snaps the spawned prefabs to the first mesh it finds (the "default" collision layer). 

Add feature to pick another layer, instead of default.

### PR DESCRIPTION
**Pull request by w-czerski**
I've added CollisionLayer to AssetConfig.
When Place On Terrain is Enabled - then a layer can be chosen to which object will be attached on spawn.
Screenshot from 2024-10-10 13-05-47
Screenshot from 2024-10-10 13-01-01

### CODE REVIEW

File: Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.h [hunk 1]
```diff
@@ -1,35 +1,36 @@
...
     //! Information about a locations to spawn an entity.
     //! This information is generated from CSV file and is used to spawn the entity.
-    //! @param m_name is the name of the spawnable entity configuration and should be identical to name in @class
-    //! SpawnableAssetConfiguration.
+    //! @param m_name is the name of the spawnable entity configuration and should be identical to name in @class SpawnableAssetConfiguration.
     class CsvSpawnableEntityInfo
     {
     public:
-        AZ_RTTI(CsvSpawnableEntityInfo, CsvSpawnableEntityInfoTypeId);
+        AZ_RTTI(CsvSpawnableEntityInfo, "{6acce47e-9cac-4a6e-8726-ea614fb2d30d}");
```
Comment (pijaro) [inline Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.h:33]: Keep the previous style of keeping the type id

File: Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.h [hunk 2]
```diff
@@ -43,13 +44,12 @@ namespace CsvSpawner::CsvSpawnerUtils
...
-    //! The class allows user to specify multiple spawnable assets for a given name and randomization parameters for position, rotation and
-    //! scale. If more then one spawnable asset is specified for a given name, then one of the spawnable assets is randomly selected for
-    //! spawning.
+    //! The class allows user to specify multiple spawnable assets for a given name and randomization parameters for position, rotation and scale.
+    //! If more then one spawnable asset is specified for a given name, then one of the spawnable assets is randomly selected for spawning.
     class CsvSpawnableAssetConfiguration
     {
     public:
-        AZ_RTTI(CsvSpawnableAssetConfiguration, CsvSpawnableAssetConfigurationTypeId);
+        AZ_RTTI(CsvSpawnableAssetConfiguration, "{d04ea4de-b4dc-4e76-b742-c1d77203bc3e}");
```
Comment (pijaro) [inline Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.h:52]: Keep the previous style of keeping the type id

File: Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.h [hunk 3]
```diff
@@ -1,35 +1,36 @@
...
 #include <AzCore/Asset/AssetCommon.h>
 #include <AzCore/Math/Transform.h>
 #include <AzCore/Math/Vector3.h>
 #include <AzCore/std/string/string.h>
+#include <AzFramework/Physics/PropertyTypes.h>
 #include <AzFramework/Spawnable/Spawnable.h>
 #include <AzFramework/Spawnable/SpawnableEntitiesInterface.h>
 
-namespace CsvSpawner::CsvSpawnerUtils
+namespace RobotecEnvironmentSpawner::CsvSpawnerUtils
```
Comment (pijaro) [inline Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.h:24]: Wrong namespace.

File: Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.h [hunk 4]
```diff
@@ -83,4 +88,4 @@ namespace CsvSpawner::CsvSpawnerUtils
         const AZStd::string& physicsSceneName = AZStd::string(),
         AZ::EntityId parentId = AZ::EntityId());
 
-}; // namespace CsvSpawner::CsvSpawnerUtils
+}; // namespace RobotecEnvironmentSpawner::CsvSpawnerUtils
```
Comment (pijaro) [inline Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.h:91]: Wrong namespace.

File: Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp [hunk 5]
```diff
@@ -238,9 +292,8 @@ namespace CsvSpawner::CsvSpawnerUtils
                 auto* transformInterface = root->FindComponent<AzFramework::TransformComponent>();
                 transformInterface->SetWorldTM(transform);
             };
-            optionalArgs.m_completionCallback =
-                [parentId](
-                    [[maybe_unused]] AzFramework::EntitySpawnTicket::Id ticketId, AzFramework::SpawnableConstEntityContainerView view)
+
+            optionalArgs.m_completionCallback = [parentId]([[maybe_unused]] AzFramework::EntitySpawnTicket::Id ticketId, AzFramework::SpawnableConstEntityContainerView view)
```
Comment (pijaro) [inline Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp:296]: Formatting

**Review by pijaro — CHANGES_REQUESTED** [review]
Please fix the namespaces and formatting. Also, some changes look like reverting to the older version (most probably because you were using the older snapshot when working on it).

**Comment by wc-robotec** [comment]
I've copied the code from version I had in Kubota repo. 
Reverted changes and applied only pieces of code I've changed.

**Review by w-czerski — COMMENTED** [review]
Updated

File: Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp [hunk 6]
```diff
@@ -161,6 +178,42 @@ namespace CsvSpawner::CsvSpawnerUtils
         return hitPosition;
     }
 
+    AZStd::optional<AZ::Vector3> RaytraceTerrain(
+        const AZ::Vector3& location, const AzPhysics::SceneHandle sceneHandle, const AZ::Vector3& gravityDirection, float maxDistance, AzPhysics::CollisionLayer collisionLayer)
+    {
+        AZStd::optional<AZ::Vector3> hitPosition = AZStd::nullopt;
+
+        if (sceneHandle == AzPhysics::InvalidSceneHandle)
```
Comment (michalpelka) [inline Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp:186]:
```suggestion
        AZ_Assert(sceneHandle == AzPhysics::InvalidSceneHandle, "Unable to get  scene handle");
        if (sceneHandle == AzPhysics::InvalidSceneHandle)
```

**Review by michalpelka — CHANGES_REQUESTED** [review]
Code is not fromated:
image
And there is a problem with missing headers
/home/michalpelka/github/ROSConDemo2023WorkDir/robotec-o3de-tools/Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.h:63:3: error: use of undeclared identifier 'AzPhysics'
                AzPhysics::CollisionLayer m_selectedCollisionLayer; //!< To which collision layer this target will be attached

**Comment by michalpelka** [comment]
Please fix compilation error:
https://gist.github.com/michalpelka/6ff222566d2a56527dc6852c13d818b4

**Comment by wc-robotec** [comment]
Fixed missing header error and added Az_Asert.
Also applied clang format - which fails ?
Could you please check if it's correctly formatted?

**Review by michalpelka — CHANGES_REQUESTED** [review]
Formatting was changed to other than o3de default. Please revert it

**Review by michalpelka — APPROVED** [review]

**Review by pijaro — APPROVED** [review]

**Pull request merged**